### ISSUE TITLE
Feature Request: `configdynamic` for static config

### ISSUE THREAD

**Issue by denysvitali** [hashicorp/waypoint#1553]
**Is your feature request related to a problem? Please describe.**

### Current situation

Currently Vault can be used as a configsource for [Dynamic Values](https://www.waypointproject.io/docs/app-config/dynamic), meaning that you're required to have the Waypoint Entrypoint to be able to dynamically fetch secrets from Vault at runtime and set them as environment variables.

Example from the website:

```hcl
config {
  env = {
    PORT = configdynamic("kubernetes", {
      name = "my-config-map" # ConfigMap name
      key  = "port"
    })
  }
}
```

This effectively means that the `configdynamic` is a placeholder for a value that is filled and automatically changed at runtime. This part is done via the Waypoint Entrypoint. As the name suggests `configdynamic` are meant for values that can and will change in the future (e.g: a password that gets rotated every X hours).
  
  
### A different use-case

Some use cases might involve injection of _static_ secrets, that is, secrets not in sync with Vault where secret rotation should not affect the workload. 

My team uses this approach to reduce the live dependency on Vault, while still having a secret store where we could put our secrets before our containers are deployed.

Currently, there is no way to inject the secrets directly as part of the waypoint.hcl entries, and thus one cannot use the secrets at (manifest) build time.

**Describe the solution you'd like**

We would like to be able to use a function similar to `configdynamic("vault", {path = "database/creds/my-role", key = "username"})` in such a way that it can be used everywhere in the `waypoint.hcl` file.

The solution should be a "compile time filler", basically a function that can be used as if it was a template variable and that can provide the secrets / values from Vault for any part of the template, similar to what we are doing with the `workspace` variable.

An example of a possible usage would be the following:

```hcl
app "app-name" {
    build {
        use "docker-pull" {
            image = somefunctionname("vault", {path = "app-name/config", key = "image-name"})
            tag   = "v1.6.1"
        }
    }
   # ...
}
```

**Describe alternatives you've considered**

A bash script to store the credentials in a file and retrieve them as environment variables (via an `envFromFile` option).

Obviously this is sub-obptimal.

**Explain any additional use-cases**

Vault might store a lot of useful variables / secrets that might be useful at build time, or deployment time. This feature will further expand the possibility to use Vault as a config source / secret source.

**Additional context**

N/A

**Comment by denysvitali** [comment]
cc/ @briancain , @evanphx , @mitchellh 

I would like to discuss the approach that I'm currently trying to use:

## waypoint-plugin-sdk

https://github.com/swisscom/waypoint-plugin-sdk/commit/e1e01308c583c60b7b0de365eb89b887be022ec8


### New `BtConfigSourcer` ("Build-time" config sourcer) in hashicorp/waypoint-plugin-sdk

```go
// BtConfigSourcer can be implemented by plugins that support sourcing
// static configuration and be used during "build-time" as variables.
//
// StaticConfigSources are meant to be used by the waypoint.hcl template
// engine and can be used even if the workload doesn't use the Waypoint
// Entrypoint.
type BtConfigSourcer interface {
	BtConfigRead() interface{}
}

// BtConfigRequest is sent to BtConfigRead for BtConfigSourcer to represent
// a request for a value from the source. Subsequent calls might returned
// cached values.
type BtConfigRequest struct {
	Name string
	Config map[string]string
}
```

## Waypoint

https://github.com/swisscom/waypoint/commit/a7f81926623c14481bc1f80c297fdfd028c12bbc

### New `btconfig()` HCL function in waypoint



```go
btConfigFunc = function.New(&function.Spec{
	Params: []function.Parameter{
		{
			Name: "from",
			Type: cty.String,
		},

		{
			Name: "config",
			Type: cty.Map(cty.String),
		},
	},
	Type: function.StaticReturnType(typeBtConfig),
	Impl: func(args []cty.Value, retType cty.Type) (cty.Value, error) {
		var config map[string]string
		if err := gocty.FromCtyValue(args[1], &config); err != nil {
			return cty.NilVal, err
		}

		// Resolve the variable here, and then return a string (now it's a BtVal)

		return cty.CapsuleVal(typeBtConfig, &pb.ConfigVar_BtVal{
			From:   args[0].AsString(),
			Config: config,
		}), nil
	},
})
```

```hcl
project = btconfig("echo", {
  value = "hello"
})
```

---

I would like to understand exactly if it makes sense to load the plugins in the `Load()` phase of the config, or if it would make sense to resolve the variable afterwards. Unfortunately most of the config entry expect a string, so passing a `BtVal` is kind of impossible.

Do you have thoughts on this approach? How would you tackle the problem?

**Comment by evanphx** [comment]
Hi @denysvitali!

Some great ideas here. We've been thinking similarly over the past couple months. What you're really looking to do is figure out how a plugin could add a function to HCL such that it could be called in various contexts.

The usage is really not dissimilar to how mappers work, interacting with plugins in their boundaries.

The question is if we were to unlock this functionality, should it be in the limited but simpler view you have here, of adding an explicit build time plugin. Or should plugins have the ability to expose arbitrary HCL functions?

**Comment by denysvitali** [comment]
> What you're really looking to do is figure out how a plugin could add a function to HCL such that it could be called in various contexts.

Not really. I'm actually using the same thing you did with `configdynamic`. That part is solved already I would say.

The biggest challenge that I see currently is that when we load the config, the plugins aren't available and thus I cannot call them to check if any plugin can satisfy the request. 

> The question is if we were to unlock this functionality, should it be in the limited but simpler view you have here, of adding an explicit build time plugin. Or should plugins have the ability to expose arbitrary HCL functions?

This is a good question. IMHO we could end up polluting the function space (e.g: clashing functions), but it would definitely be awesome to be able to do something like:

```hcl
name = vault("some/path", "xyz")
```

**Comment by mitchellh** [comment]
More specifically to config sourcers, we've also recently discussed making config sourcers work in the runner environment and adding a variable type to reference it (i.e. `config.my_vault_var` -- this is purely hypothetical). Then this would "just work".

This comes with various downsides: the runner needs access to Vault (both network and auth), we have to load plugins in the runner (security concerns and would have to get the plugins there). 

One of the things @evanphx is working on for the next major release is a concept we call "ad hoc runners" which we launch just-in-time using any container image. This might make this considerably easier and more attractive for this functionality...

**Comment by denysvitali** [comment]
> the runner needs access to Vault

I honestly don't see it as a proper drawback, but rather as a prerequisite. In our case, the connection to Vault from the runner is already established, and the Vault login credentials are already in the config parameters (which are used already for `configdynamic`)

> we have to load plugins in the runner

This is in fact true, and we need to find the best way to do it. We could follow a similar approach as `configdynamic`, but in this case we should find a way to deal with the type system since we will be returning another type instead of a string. For this reason:

```hcl

name = some_func("vault", {})
```

Might cause a type error if some_func doesn't immediately get validated as a string.

This effectively causes the following choices:

- We can "build" the manifest before it's even interpreted by Waypoint's `ReadConfig`, basically solving the type issue
- We find a way to allow the `DynamicValue` type and interpret its value later (after ReadConfig). This will probably require us to mess up the whole config structure / parser but at least will give us a way to execute the functions at a later stage, effectively making it easier to load the plugins dynamically

**Comment by evanphx** [comment]
It probably wouldn't work exactly like `configdynamic` rather it would just return the value as a `string`. `configdynamic` returns a unique type because it's actually communicating the desire to perform a function in the future (in the entrypoint), where as this would be a case of needing to just perform the function at invocation time.

**Comment by izaaklauer** [comment]
https://github.com/hashicorp/waypoint/pull/2889 provides one solution to this use-case - for any platform plugin with a static environment field ([i.e. ecs](https://www.waypointproject.io/plugins/aws-ecs#static_environment)), it will be possible to use `configdynamic` to populate static entries there at deploy-time.

I'd also note this approach has security ramifications - when the CEB gets its secrets from vault at runtime, the secret is exposed only to the app. With this method, the secret will be visible to the waypoint runner, when it renders the config. It will also be exposed to the cloud platform - with ecs for example, it will be visible in the aws console.

**Comment by mitchellh** [comment]
I believe this is resolved with #2889
